xds/server.go

```diff
@@ -232,22 +205,17 @@ func (s *GRPCServer) newListenerWrapper(lis net.Listener) (*listenerWrapper, err
 	}
 	name := fmt.Sprintf("%s?udpa.resource.listening_address=%s", path, lis.Addr().String())
 
-	// Register an LDS watch using our xdsClient, and specify the listening
-	// address as the resource name.
-	cancelWatch := s.xdsC.WatchListener(name, func(update xdsclient.ListenerUpdate, err error) {
-		s.handleListenerUpdate(listenerUpdate{
-			lw:         lw,
-			name:       name,
-			lds:        update,
-			err:        err,
-			goodUpdate: goodUpdate,
-		})
+	// Create a listenerWrapper which handles all functionality required by
+	// this particular instance of Serve().
+	lw, goodUpdateCh := server.NewListenerWrapper(server.ListenerWrapperParams{
+		Listener:             lis,
+		Logger:               s.logger,
+		ListenerResourceName: name,
+		XDSCredsInUse:        s.xdsCredsInUse,
+		CertProviderConfigs:  certProviderConfigs,
+		WatchFunc:            s.xdsC.WatchListener,
+		BuildProviderFunc:    buildProvider,
 	})
```

Comment:
This is a very wide API.  Is it possible to pare it down?

I don't think we need to pass the logger.  You should be able to recreate it in the other package and use it directly that way.

It seems like most of these parameters are needed for the implementation of `XDSHandshakeInfo()` on the conn - would it make sense to inject all or part of that implementation as a closure maybe?  If it's too hard or doesn't make sense to tear that out, that's fine.

Reply:
Yes, much better, thanks.